apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1/HiPerfPublisher.cs

```diff
@@ -0,0 +1,626 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using Google.Api.Gax;
+using Google.Api.Gax.Grpc;
+using Google.Apis.Auth.OAuth2;
+using Google.Cloud.PubSub.V1.Tasks;
+using Google.Protobuf;
+using Grpc.Auth;
+using Grpc.Core;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Google.Cloud.PubSub.V1
+{
+    /// <summary>
+    /// A PubSub publisher that is associated with a specific <see cref="TopicName"/>.
+    /// </summary>
+    public abstract class HiPerfPublisher
+    {
+        // TODO: Logging
+
+        /// <summary>
+        /// Settings for the <see cref="HiPerfPublisher"/>.
+        /// </summary>
+        public sealed class Settings
+        {
+            /// <summary>
+            /// Instantiate settings with pre-created <see cref="PublisherClient"/>s.
+            /// </summary>
+            /// <param name="clients">The <see cref="PublisherClient"/>s to use in a <see cref="HiPerfPublisher"/>"/>.</param>
+            public Settings(IList<PublisherClient> clients) => Clients = clients;
+
+            /// <summary>
+            /// Instantiate settings, optionally specifying client creation parameters.
+            /// </summary>
+            /// <param name="clientCount">Optional. The number of clients to create.
+            /// Defaults to the CPU count on the machine this is being executed on.</param>
+            /// <param name="publisherSettings">Optional. The settings to use when creating <see cref="Publisher"/> instances.</param>
+            /// <param name="credentials">Optional. Credentials to use then creating <see cref="Publisher"/> instances.
+            /// Defaults to using the default credentials.</param>
+            public Settings(int? clientCount = null, PublisherSettings publisherSettings = null, ChannelCredentials credentials = null)
+            {
+                ClientCount = clientCount;
+                PublisherSettings = publisherSettings;
+                Credentials = credentials;
+            }
+
+            internal Settings(Settings other)
+            {
+                Clients = other.Clients != null ? new List<PublisherClient>(other.Clients) : null;
+                ClientCount = other.ClientCount;
+                PublisherSettings = other.PublisherSettings?.Clone();
+                Credentials = other.Credentials;
+                BatchingSettings = other.BatchingSettings?.Clone();
+                Scheduler = other.Scheduler;
+                BatchMaxElementCount = other.BatchMaxElementCount;
+                BatchMaxByteCount = other.BatchMaxByteCount;
+            }
+
+            /// <summary>
+            /// The clients to use within a <see cref="HiPerfPublisher"/>.
+            /// </summary>
+            public IList<PublisherClient> Clients { get; }
```

Comment:
Thinking about it again, it feels odd for Clients to be part of the settings in the first place. It doesn't feel like the other settings at all...